src/python/pim/polynomial_representation.py

```diff
@@ -0,0 +1,89 @@
+from typing import List
+import re
+import itertools as it
+import numpy as np
+
+class Polynomial:
+    def __init__(self, coefficients: List[int]) -> None:
+        self.coefficients: List[int] = coefficients
+
+    def __str__(self) -> str:
+        """returns a string representation of the polynomial"""
+        coeff_list: List[str] = [display(c, i) for i, c in enumerate(self.coefficients)]
+        filter_coeff_list = list(filter(lambda x: x != "", coeff_list))
+        return re.sub(r"\+ -", "- ", " + ".join(filter_coeff_list))
+
+    def __add__(self, other: "Polynomial") -> "Polynomial":
+        """returns the sum of two polynomials (like degrees are added together)"""
+        return [sum(n) for n in it.zip_longest(self.coefficients, other.coefficients, fillvalue = 0)]
+
+    def __mul__(self, other: "Polynomial") -> "Polynomial":
+        """returns the product of two polynomials"""
+        len_self: int = len(self.coefficients)
+        len_other: int = len(other.coefficients)
+        blank_coeff_list: list(int) = [0] * (len_self + len_other - 1)
+
+        #Multiplies two polynomials term by term
+        for i in range(len_self):
+            for j in range(len_other):
+                blank_coeff_list[i + j] += self.coefficients[i] * other.coefficients[j]
+        return blank_coeff_list
+
+    def evaluate(self, value: int) -> int:
+        """evaluates a polynomial at a given value"""
+        calculate_list: list[str] = [calculate(c, i) for i, c in enumerate(self.coefficients)]
+        total: int = 0
+        for i in range(len(self.coefficients)):
+            total += calculate_list[i][0] * (value ** calculate_list[i][1])
+        return total
+
+def display(coefficient: int, index: int) -> str:
+    if index == 0:
+        x: str = ""
+    elif index == 1:
+        x: str = "x"
+    else:
+        x: str = "x^"+str(index)
+    if coefficient == 0:
+        output: str = ""
+    else:
+        output: str = str(coefficient) + x
+
+    return output
+
+def calculate(coefficient: int, index: int) -> (int, int):
```

Comment:
You are returning a `Tuple[int, int]`, but the concept is a pairing of coefficient and order. That is, you are returning a single polynomial term. One thing, you could do for clarity (which I often do) is create a very simple [`namedtuple`](https://docs.python.org/3/library/collections.html#collections.namedtuple). 

```python
PolyTerm = namedtuple("PolyTerm", ["coefficient", "order"])
```

So this would imply, that body of your function would look something like the following:

```python
def calculate(coefficient: int, index: int) -> PolyTerm:
    if coefficient == 0:
        return PolyTerm(0, 0)
    else:
        return PolyTerm(coefficient, index)
```

Once you have have an instance of `PolyTerm`, you can reference things directly by name:

```python
pt: PolyTerm = PolyTerm(1, 2)
assert pt.coefficient == 1
assert pt.order == 2
```

This is not material from a mechanical perspective, but might make it easier to read the code.

Reply:
I adjusted the calculate and evaluate functions to use a `PolyTerm` _namedtuple_; however I have not figured out yet how to actually make it realize I have defined `PolyTerm`. (I get the error of `name 'PolyTerm' is not defined` when I try to run the test code.

Reply:
I am guessing that if you move the definition of `PolyTerm` in front of `Polynomial` (where you use it), you probably don't have to put it in quotes.